### PR TITLE
[AutoDiff upstream] Add `@transpose` attribute.

### PR DESCRIPTION
**Pull request by saeta**
The `@transpose(of:)` attribute registers a function as a transpose of another
function. This patch adds the `@transpose(of:)` attribute definition, syntax,
parsing, and printing.

Resolves TF-827.

---

Examples:

```swift
extension FloatingPoint where Self: Differentiable {
    @transpose(of: +)
    static func transposeAdd(_ v: Self) -> (Self, Self) { (v, v) }

    @transpose(of: *, wrt: 0)
    @transpose(of: *, wrt: 1)
    static func transposeMultiply(lhs: Self, rhs: Self) -> Self { lhs * rhs }
}
```

The [differentiable programming manifesto](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md#make-a-function-differentiable-using-derivative-or-transpose) has more information and examples.

---

Note: this code is upstreamed from `tensorflow` branch.

Known todos and issues are referenced in comments:
- Upstream `@transpose` attribute type-checking (TF-830, TF-1060).
- Enable `@transpose` attribute serialization (TF-838).
- Use module-qualified names instead of custom qualified name syntax/parsing
  (TF-1066).

We would like to upstream code and continue development on `master`, if that's acceptable!

### CODE REVIEW

**Comment by dan-zheng** [comment]
I revived this PR to upstream the current state of `tensorflow` branch, it is now ready for review. cc @rxwei @lattner 

The PR description tracks todos and known issues.

File: include/swift/AST/Attr.h [hunk 1]
```diff
@@ -1858,6 +1858,89 @@ class DerivativeAttr final
   }
 };
 
+/// The `@transpose` attribute registers a function as a transpose of another
```
Comment (lattner) [inline include/swift/AST/Attr.h]: nit: `@transpose(of:)`

File: test/AutoDiff/Parse/transpose_attr_parse.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+@transpose(of: foo, wrt: (self, 0, 1, 2))
+func transpose(v: Float) -> (Float, Float, Float, Float)
+
+// Qualified declaration.
+@transpose(of: A.B.C.foo(x:y:_:z:))
+func transpose(v: Float) -> Float
+
+// Qualified operator.
+// TODO(TF-1065): Consider disallowing qualified operators.
```
Comment (lattner) [inline test/AutoDiff/Parse/transpose_attr_parse.swift]: FWIW, I think that your implemented behavior is correct and nothing needs to change here.  "TF-1065" is really about the existing swift compiler being buggy and not allowing qualified names for operators like ..< . I'd suggest splitting that off to a swift bug, and I'd recommend that operator having to be spelled as Swift.`..<`

**Review by lattner — APPROVED** [review]
Nice!

**Review by rxwei — APPROVED** [review]

File: test/AutoDiff/Parse/transpose_attr_parse.swift (hunk 2 above)
Comment (dan-zheng) [reply test/AutoDiff/Parse/transpose_attr_parse.swift]: Thanks for the thorough review! I started a [discussion](https://forums.swift.org/t/operator-member-syntax/31773) regarding operator member syntax with proposed backtick syntax (`` SIMD2<Float>.`.<` ``).

As stated in TF-1065, qualified operator syntax isn't strictly necessary for `@transpose` or `@derivative` attribute because type-checking for these attributes require the operator to be defined in the same type context, where lookup is not ambiguous.

Taking action on these parser changes seems low priority so I'll leave the code and tests as is for now.

File: include/swift/AST/Attr.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/Attr.h]: Done in https://github.com/apple/swift/pull/27545/commits/b94c097de935fd9a93b358aaf2d17f52c5e0b748.

**Comment by dan-zheng** [comment]
https://github.com/apple/swift-syntax/pull/189
@swift-ci Please test

**Comment by dan-zheng** [comment]
CI failed for `test/IDE/complete_decl_attribute.swift` but the test passes for me locally.
Perhaps a clean test is necessary, I'll try it now.

**Comment by dan-zheng** [comment]
apple/swift-syntax#189
@swift-ci Please clean test

File: lib/Parse/ParseType.cpp [hunk 3]
```diff
@@ -633,12 +633,42 @@ ParserStatus Parser::parseGenericArguments(SmallVectorImpl<TypeRepr *> &Args,
...
 }
 
+/// Returns true if a base type for a qualified declaration name can be
+/// parsed.
+///
+/// Examples:
+///   'Foo.f' -> true
+///   'Foo.Bar.f' -> true
+///   'f' -> false, no base type
+bool Parser::canParseBaseTypeForQualifiedDeclName() {
```
Comment (rintaro) [inline lib/Parse/ParseType.cpp]: It would be great if  this can share some code with `canParseTypeIdentifier()`.
Also please move this function below line 1374.

File: test/AutoDiff/Parse/transpose_attr_parse.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+@transpose(of: Swift.Float..<)
+func transpose(v: Float) -> Float
+
+// `init` keyword edge case.
+@transpose(of: Swift.Float.init(_:))
+func transpose(v: Float) -> Float
+
+// `subscript` keyword edge case.
+@transpose(of: Swift.Array.subscript(_:))
```
Comment (rintaro) [inline test/AutoDiff/Parse/transpose_attr_parse.swift]: Could you add test cases for specialized generic types? Like: `@transpose(of: Foo<Int>.Bar.funcname)`

File: utils/gyb_syntax_support/AttributeNodes.py [hunk 5]
```diff
@@ -311,13 +314,56 @@
...
+         An optionally qualified function declaration name (e.g. `+(_:_:)`,
+         `A.B.C.foo(_:_:)`).
+         ''',
+         children=[
+             Child('BaseType', kind='Type', description='''
+                   The base type of the qualified name, optionally specified.
+                   ''',
+                   node_choices=[
+                       Child('MemberType', kind='MemberTypeIdentifier'),
+                       Child('SimpleType', kind='SimpleTypeIdentifier'),
```
Comment (rintaro) [inline utils/gyb_syntax_support/AttributeNodes.py]: I know the parser doesn't allow `@traspose(of: [Int].somehting)` But as a syntax declaration, I don't think need to limit this for type identifiers (i.e. we don't need `node_choices` at all).

File: utils/gyb_syntax_support/AttributeNodes.py [hunk 6]
```diff
@@ -311,13 +314,56 @@
...
+             Child('Name', kind='Syntax', description='''
+                   The base name of the referenced function.
+                   ''',
+                   node_choices=[
+                       Child('Identifier', kind='IdentifierToken'),
+                       Child('PrefixOperator', kind='PrefixOperatorToken'),
+                       Child('PostfixOperator', kind='PostfixOperatorToken'),
+                       Child('SpacedBinaryOperator',
+                             kind='SpacedBinaryOperatorToken'),
+                   ]),
```
Comment (rintaro) [inline utils/gyb_syntax_support/AttributeNodes.py]:
```python
    kind='Token',
    token_choices=[
        'IdentifierToken',
        'UnspacedBinaryOperatorToken',
        'SpacedBinaryOperatorToken',
        'PrefixOperatorToken',
        'PostfixOperatorToken',
    ]),
```
This should work. e.g. `FunctionDeclSyntax.Identifier`

Comment (dan-zheng) [reply utils/gyb_syntax_support/AttributeNodes.py]: Done in https://github.com/apple/swift/pull/27545/commits/c736938911f0d2c4b5fd08968c365bc58e1c0539, thank you!

Comment (dan-zheng) [reply utils/gyb_syntax_support/AttributeNodes.py]: Would you like to be added as a reviewer on future parsing/syntax PRs? I'm happy to do so!

File: utils/gyb_syntax_support/AttributeNodes.py (hunk 5 above)
Comment (dan-zheng) [reply utils/gyb_syntax_support/AttributeNodes.py]: Done in https://github.com/apple/swift/pull/27545/commits/c736938911f0d2c4b5fd08968c365bc58e1c0539.

File: test/AutoDiff/Parse/transpose_attr_parse.swift (hunk 4 above)
Comment (dan-zheng) [reply test/AutoDiff/Parse/transpose_attr_parse.swift]: Specialized generic type test added in https://github.com/apple/swift/pull/27545/commits/c736938911f0d2c4b5fd08968c365bc58e1c0539. It passes, as expected.

File: lib/Parse/ParseType.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/Parse/ParseType.cpp]: Thanks for the catch!

I moved `Parser::canParseBaseTypeForQualifiedDeclName` just below `Parser::canParseTypeIdentifier` to line 1562. I also removed the doc comment on the function definition to be consistent with the surrounding code.

Comment (dan-zheng) [reply lib/Parse/ParseType.cpp]: Code sharing done in https://github.com/apple/swift/pull/27545/commits/fa31c75aac4ac56c3b15e59b42bdc0c941cb2498.

I factored code into a new utility `Parser::canParseSimpleTypeIdentifier`, used by both `Parser::canParseTypeIdentifier` and `Parser::canParseBaseTypeForQualifiedDeclName`.

File: utils/gyb_syntax_support/AttributeNodes.py (hunk 6 above)
Comment (rintaro) [reply utils/gyb_syntax_support/AttributeNodes.py]: > Would you like to be added as a reviewer on future parsing/syntax PRs?

Yes. I appreciate it!

**Review by rintaro — APPROVED** [review]
Parser/Syntax changes look good!

File: utils/gyb_syntax_support/AttributeNodes.py (hunk 6 above)
Comment (rintaro) [reply utils/gyb_syntax_support/AttributeNodes.py]: > Would you like to be added as a reviewer on future parsing/syntax PRs?

Yes. I appreciate it

Comment (dan-zheng) [reply utils/gyb_syntax_support/AttributeNodes.py]: Great, will do 🙂

**Comment by dan-zheng** [comment]
https://github.com/apple/swift-syntax/pull/189
@swift-ci Please clean test
